File: src/tiles.rs
//! XYZ MVT extension. Spatial filtering is CRS84; tile encoding is Web Mercator.
use crate::{
    api::{self, SourceQuery},
    db, filter,
    store::{Error, Store},
};
use bytes::Bytes;
use poem::{
    handler,
    http::StatusCode,
    web::{Data, Path, Query},
    Request, Response,
};
use std::sync::Arc;

#[handler]
pub async fn tile(
    Path((collection, z, x, y)): Path<(String, u8, u32, u32)>,
    Query(query): Query<SourceQuery>,
    req: &Request,
    Data(store): Data<&Arc<Store>>,
) -> Result<Response, Error> {
    store.collection(&collection)?;
    if z > 30 || x >= (1u32 << z) || y >= (1u32 << z) {
        return Err(Error::Invalid("tile coordinates outside XYZ matrix".into()));
    }
    let sources = api::source_ids(req, query.sources.as_deref())?;
    let bbox = martin_tile_utils::xyz_to_bbox(z, x, y, x, y);
    let half = std::f64::consts::PI * 6_378_137.0;
    let span = 2.0 * half / f64::from(1u32 << z);
    let west = -half + f64::from(x) * span;
    let north = half - f64::from(y) * span;
    // SQL builds per request; validation runs before any database work.
    // Tiles hold up to 5k features and run Mercator transforms, so they
    // share the bulk lane with Flight.
    let from = store.read_source(Some(bbox));
    let body = store
        .run_bytes(true, move |conn| {
            // Single scan: the HAVING maps empty tiles to zero rows
            // (204) instead of an empty MVT blob (200).
            let fetch = Store::predicate(&collection, Some(bbox), &sources);
            // Page first, transform second: the inner scan selects raw
            // id/geom for the page, the outer runs Mercator + clip only
            // over those rows instead of every scanned match.
            let sql = format!(
                "SELECT ST_AsMVT(t, {}) FROM (SELECT id, ST_AsMVTGeom(\
                 ST_Transform(page.geom, 'EPSG:4326', 'EPSG:3857', always_xy := true), \
                 ST_Extent(ST_MakeEnvelope({west}, {}, {}, {north})), 4096, 64, true) AS geom \
                 FROM (SELECT id, geom FROM {from} WHERE {fetch} ORDER BY id LIMIT 5000) AS page) t \
                 WHERE geom IS NOT NULL AND NOT ST_IsEmpty(geom) HAVING count(*) > 0",
                filter::quote(&collection),
                north - span,
                west + span
            );
            let tile = db::blob_one(conn, &sql)?.unwrap_or_default();
            Ok(Bytes::from(tile))
        })
        .await?;
    if body.bytes.is_empty() {
        Ok(Response::builder()
            .status(StatusCode::NO_CONTENT)
            .header(poem::http::header::CACHE_CONTROL, "public, max-age=60")
            .header(
                poem::http::header::VARY,
                "Accept, Accept-Encoding, X-Source-Ids",
            )
            .finish())
    } else {
        Ok(api::conditional_response(
            req,
            &body,
            "application/vnd.mapbox-vector-tile",
        ))
    }
}

File: internal/tiles/tiles.go
// Package tiles ports src/tiles.rs: XYZ MVT extension math + SQL.
package tiles

import (
	"fmt"
	"math"

	"github.com/adonm/lakewing/internal/filter"
)

// XYZToBBox converts a tile to a CRS84 bbox [w,s,e,n].
func XYZToBBox(z uint8, x, y uint32) [4]float64 {
	n := float64(uint64(1) << z)
	west := float64(x)/n*360 - 180
	east := float64(x+1)/n*360 - 180
	north := mercatorToLat(math.Pi * (1 - 2*float64(y)/n))
	south := mercatorToLat(math.Pi * (1 - 2*float64(y+1)/n))
	return [4]float64{west, south, east, north}
}

func mercatorToLat(m float64) float64 {
	return 180 / math.Pi * (2*math.Atan(math.Exp(m)) - math.Pi/2)
}

// ValidateTile mirrors the XYZ matrix check in tiles::tile.
func ValidateTile(z uint8, x, y uint32) error {
	if z > 30 || x >= (uint32(1)<<z) || y >= (uint32(1)<<z) {
		return fmt.Errorf("tile coordinates outside XYZ matrix")
	}
	return nil
}

// MVTSQL mirrors the single-scan MVT query in src/tiles.rs.
func MVTSQL(collection, from, fetch string, west, south, east, north float64) string {
	return fmt.Sprintf(
		"SELECT ST_AsMVT(t, %s) FROM (SELECT id, ST_AsMVTGeom("+
			"ST_Transform(page.geom, 'EPSG:4326', 'EPSG:3857', always_xy := true), "+
			"ST_Extent(ST_MakeEnvelope(%v, %v, %v, %v)), 4096, 64, true) AS geom "+
			"FROM (SELECT id, geom FROM %s WHERE %s ORDER BY id LIMIT 5000) AS page) t "+
			"WHERE geom IS NOT NULL AND NOT ST_IsEmpty(geom) HAVING count(*) > 0",
		filter.Quote(collection), west, south, east, north, from, fetch)
}
